Healchecks via https are succefull in case of certificate issue

I discovered strange vault injector behavior. I use liveness probe like:
```
Liveness:   http-get https://:8080/health/ready delay=30s timeout=30s period=30s #success=1 #failure=3
```
After some very strage events in may k8s cluster, vault injector failed 3 probes and was restarted by kubelet:
```
 [ERROR] handler: http: TLS handshake error from 10.0.216.118:51410: EOF
 [ERROR] handler: http: TLS handshake error from 10.0.216.118:37002: EOF
 [ERROR] handler: http: TLS handshake error from 10.0.216.118:55394: EOF
```
It is still normal and expected behavior. But after restart vault injector could not read certificate from filesystem with endless logs like:
```
handler.certwatcher: Could not load TLS keypair: tls: failed to find any PEM data in certificate input. Trying again...
handler.certwatcher: Could not load TLS keypair: tls: failed to find any PEM data in certificate input. Trying again...
...
```
I have no idea how it can be possible. I can read cert and key file inside vault injector container without problem. May be some memory issue in k8s node...
In any way, tls in vault injector is broken, but `curl https://pod_address:8080/health/ready` return 204 code as correctly working instance.
If healthcheck request fail in this scenario, kubelet will restart container and vault injector starts working normally.

May be add some try threshold for tls issue? And after threshold's excess health check endpoint will response error code.

Application deployment:


**Environment**
* Kubernetes version: 1.20.11
  * Baremetal
* vault-k8s version: 0.14.0

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,7 +1,12 @@
 ## Unreleased
 
+Improvements:
+* ConfigMap with missing vault section should default to env vars [GH-353](https://github.com/hashicorp/vault-k8s/pull/353)
+
 Changes:
 * Certificate watcher timer deadlock fix [GH-350]([https://github.com/hashicorp/vault-k8s/pull/350)
+* Wait for certificate before starting HTTP listener [GH-354](https://github.com/hashicorp/vault-k8s/pull/354)
+* Update example injector mutating webhook config to exclude agent pod [GH-351](https://github.com/hashicorp/vault-k8s/pull/351)
 
 ## 0.16.0 (May 11, 2022)
 

--- a/helper/cert/notify.go
+++ b/helper/cert/notify.go
@@ -2,17 +2,19 @@ package cert
 
 import (
 	"context"
+	"sync"
 	"time"
 
 	"github.com/hashicorp/go-hclog"
 )
 
-func NewNotify(ctx context.Context, newBundle chan<- Bundle, source Source, logger hclog.Logger) *Notify {
+func NewNotify(ctx context.Context, newBundle chan<- Bundle, notifyOnce chan<- bool, source Source, logger hclog.Logger) *Notify {
 	return &Notify{
-		ctx:    ctx,
-		ch:     newBundle,
-		source: source,
-		logger: logger,
+		ctx:        ctx,
+		ch:         newBundle,
+		notifyOnce: notifyOnce,
+		source:     source,
+		logger:     logger,
 	}
 }
 
@@ -26,6 +28,8 @@ type Notify struct {
 	// blocks then the notify loop will also be blocked, so downstream
 	// users should process this channel in a timely manner.
 	ch chan<- Bundle
+	// used to notify the first time it is run
+	notifyOnce chan<- bool
 
 	// source is the source of certificates.
 	source Source
@@ -39,6 +43,7 @@ func (n *Notify) Run() {
 	retryTicker := time.NewTicker(5 * time.Second)
 	defer retryTicker.Stop()
 	first := true
+	once := sync.Once{}
 	for {
 		if first {
 			// On the first pass we want to check for the cert immediately.
@@ -57,6 +62,12 @@ func (n *Notify) Run() {
 			n.logger.Warn("error loading next cert", "error", err.Error())
 			continue
 		}
+
+		once.Do(func() {
+			go func() {
+				n.notifyOnce <- true
+			}()
+		})
 
 		// If the returned bundles are equal, then ignore it.
 		if last.Equal(&next) {

--- a/helper/cert/notify_test.go
+++ b/helper/cert/notify_test.go
@@ -19,7 +19,8 @@ func TestNotify(t *testing.T) {
 
 	// Create notifier
 	ch := make(chan Bundle)
-	n := NewNotify(context.Background(), ch, source, hclog.NewNullLogger())
+	notifyOnce := make(chan bool)
+	n := NewNotify(context.Background(), ch, notifyOnce, source, hclog.NewNullLogger())
 	go n.Run()
 
 	// We should receive an update almost immediately
@@ -35,6 +36,19 @@ func TestNotify(t *testing.T) {
 	case <-time.After(750 * time.Millisecond):
 	case <-ch:
 		t.Fatal("should not receive update")
+	}
+
+	// we should receive the once notification
+	select {
+	case <-time.After(250 * time.Millisecond):
+		t.Fatal("should've received once notification")
+	case <-notifyOnce:
+	}
+	// ... just once
+	select {
+	case <-time.After(250 * time.Millisecond):
+	case <-notifyOnce:
+		t.Fatal("should not receive second notification on notifyOnce channel")
 	}
 
 	b := <-ch
@@ -54,6 +68,8 @@ func TestNotifyRace(t *testing.T) {
 	// giving races a chance to be detected.
 	done := make(chan interface{})
 
+	notifyOnce := make(chan bool)
+
 	for i := 0; i < numParallel; i++ {
 
 		// Use one background context for everything since that could happen
@@ -68,7 +84,7 @@ func TestNotifyRace(t *testing.T) {
 			Hosts: []string{"some", "hosts"},
 			Log:   hclog.Default(),
 		}
-		n := NewNotify(ctx, certCh, certSource, hclog.NewNullLogger())
+		n := NewNotify(ctx, certCh, notifyOnce, certSource, hclog.NewNullLogger())
 
 		go func() {
 			<-start

--- a/subcommand/injector/command.go
+++ b/subcommand/injector/command.go
@@ -176,7 +176,8 @@ func (c *Command) Run(args []string) int {
 	// Create the certificate notifier so we can update for certificates,
 	// then start all the background routines for updating certificates.
 	certCh := make(chan cert.Bundle)
-	certNotify := cert.NewNotify(ctx, certCh, certSource, logger.Named("notify"))
+	certReady := make(chan bool)
+	certNotify := cert.NewNotify(ctx, certCh, certReady, certSource, logger.Named("notify"))
 	go certNotify.Run()
 	go c.certWatcher(ctx, certCh, clientset, leaderElector, logger.Named("certwatcher"))
 
@@ -222,6 +223,18 @@ func (c *Command) Run(args []string) int {
 		c.UI.Error(fmt.Sprintf("Failed to configure TLS: %s", err))
 		return 1
 	}
+
+	// if we don't have a certificate given to us, wait for cert to generate to prevent
+	// TLS errors on startup
+	if c.flagCertFile == "" && c.flagKeyFile == "" {
+		<-certReady
+	} else {
+		// otherwise drain async just to clean up the sending goroutine
+		go func() {
+			<-certReady
+		}()
+	}
+
 	server := &http.Server{
 		Addr:      c.flagListen,
 		Handler:   handler,

